neural_speed/convert/convert_stablelm.py

```diff
@@ -223,45 +146,72 @@ def stablelm_convert(model, tokenizer, dir_model, fname_out, ftype, hparams):
             fout.write(text)
             fout.write(struct.pack("f", -10000))
 
-    list_vars = model.state_dict()
+    def write_header(name, data, ftype=0):
+        str = name.encode('utf-8')
+        n_dims = len(data.shape)
+        fout.write(struct.pack("iii", n_dims, len(str), ftype))
+        for i in range(n_dims):
+            fout.write(struct.pack("i", data.shape[n_dims - 1 - i]))
+        print(str)
+        fout.write(str)
 
     print(hparams)
+    q_norms, k_norms = dict(), dict()
+    for name, data_torch in list_vars.items():
+        # Convert any unsupported data types to float32
+        if data_torch.dtype not in (torch.float16, torch.float32):
+            data_torch = data_torch.to(torch.float32)
+
+        # Skip some tensors
+        if name.endswith((".attention.rotary_emb.inv_freq")):
+            continue
 
-    for name in list_vars.keys():
-        # No gradients for these
-        list_vars[name].requires_grad = False
-        src = name
-        print(src, ' -> ', name)
-        data = list_vars[src].squeeze().numpy()
-        data = data.astype(np.float32)
-
+        data = data_torch.squeeze().numpy()
+        old_dtype = data.dtype
         n_dims = len(data.shape)
-        print(name, n_dims, data.shape)
+        if name.find("q_layernorm.norms") != -1:
```

Comment:
could you please explain these codes? I'm a bit confused about the `qk_norm` stacking way.

Reply:
Yes, it uses per head qk_normalization, we basically just stack all normalization weights across the query heads + across the the key heads, and then apply normalization at once. Check also discussion in llama.cpp PR since they did it in same way: https://github.com/ggerganov/llama.cpp/pull/6635